mips64r6el Cross Toolchain: new

Cyanoxygen opened the pull request:
Topic Description
-----------------

This topic brings MIPS64 R6 cross build toolchain.

Package(s) Affected
-------------------

- `binutils+cross-mips64r6el`
- `gcc+cross-mips64r6el`

Security Update?
----------------

No

Test Build(s) Done
------------------

**Primary Architectures**

<!-- Please remove any architecture to which this topic does not apply. -->

- [x] AMD64 `amd64`   
- [ ] AArch64 `arm64`

**Secondary Architectures**

Architectural progress for "secondary," or experimental ports does not impede on merging of this topic.

- [ ] Loongson 3 `loongson3`
- [ ] MIPS R6 64-bit (Little Endian) `mips64r6el`
- [ ] PowerPC 64-bit (Little Endian) `ppc64el`
- [ ] RISC-V 64-bit `riscv64`

Update(s) Uploaded to Stable
----------------------------

**Primary Architectures**

- [ ] AMD64 `amd64`   
- [ ] AArch64 `arm64`

**Secondary Architectures**

Architectural progress for "secondary," or experimental ports does not impede on merging of this topic.

- [ ] Loongson 3 `loongson3`
- [ ] MIPS R6 64-bit (Little Endian) `mips64r6el`
- [ ] PowerPC 64-bit (Little Endian) `ppc64el`
- [ ] RISC-V 64-bit `riscv64`

Cyanoxygen committed:
binutils+cross-mips64r6el: new, 2.40

Cyanoxygen committed:
gcc+cross-mips64r6el: new, 12.2.0

cthbleachbit commented on app-devel/binutils+cross-mips64r6el/autobuild/build:
```diff
@@ -0,0 +1,11 @@
+
+mkdir -p build
+cd build
+../configure --prefix=/opt/abcross/mips64r6el --target=mipsisa64r6el-aosc-linux-gnuabi64 \
```
Nit: Split each parameter on a separate line

cthbleachbit commented on app-devel/gcc+cross-mips64r6el/autobuild/build:
```diff
@@ -0,0 +1,14 @@
+
+mkdir -p build
+cd build
+AR=ar ../configure --prefix=/opt/abcross/mips64r6el --target=mipsisa64r6el-aosc-linux-gnuabi64 \
```
- Split parameters into separate lines.
- Why we need an `AR=ar` here?

cthbleachbit commented on app-devel/gcc+cross-mips64r6el/autobuild/build:
```diff
@@ -0,0 +1,14 @@
+AR=ar ../configure --prefix=/opt/abcross/mips64r6el --target=mipsisa64r6el-aosc-linux-gnuabi64 \
+	--with-sysroot=/var/ab/cross-root/mips64r6el --enable-shared --disable-multilib \
+	--enable-c99 --enable-long-long --enable-threads=posix --enable-languages=c,c++,fortran,lto \
+	--enable-__cxa_atexit --disable-altivec --disable-fixed-point --with-arch=mips64r6 --with-tune=mips64r6 \
```
Altivec is only valid for power architecture and is irrelavant for mips64r6. Please remove `--disable-altivec`

Cyanoxygen committed:
binutils+cross-mips64r6el: lint build script

- Split configure options into separate lines.
- Add respective abinfo output before each steps.

Co-Authored-By: Thanhao Chai <cth451@gmail.com>

Cyanoxygen committed:
gcc+cross-mips64r6el: Switch to minimal sysroot

- This greatly reduces required sources to download.
- gcc is able to build simple program with this minimal sysroot shipped
  with the toolchain. Before that, users have to prepare an environment
  by themselves.
- A complete OS environment is still required if someone want to cross
  build with existing libraries.

Cyanoxygen committed:
gcc+cross-mips64r6el: extract minimal package set ...
instead of complete BuildKit

Cyanoxygen committed:
gcc+cross-mips64r6el: add respective abinfo output

- Add respective abinfo output before each steps.

Cyanoxygen committed:
gcc+cross-mips64r6el: lint build script

- Split configure arguments into separate lines.

Co-Authored-By: Tianhao Chai <cth451@gmail.com>

Cyanoxygen committed:
gcc+cross-mips64r6el: copy minimal sysroot to PKGDIR

- A minimal sysroot is required to actually build simple programs, i.e.
  program written with no dependency other than glibc.

Cyanoxygen commented:
Testing
--------

This toolchain is able to build simple program, i.e. program written with only C library functions.

![image](https://user-images.githubusercontent.com/29174191/227121555-34b0b8f9-8f98-44fe-aa88-74e89b417a54.png)

MingcongBai commented on app-devel/binutils+cross-mips64r6el/autobuild/beyond:
```diff
@@ -0,0 +1 @@
+rm $PKGDIR/opt/abcross/mips64r6el/share/info/dir
```
Missing verbosity, quote $PKGDIR.

MingcongBai commented on app-devel/binutils+cross-mips64r6el/autobuild/build:
```diff
@@ -0,0 +1,23 @@
+unset CFLAGS CXXFLAGS CPPFLAGS LDFLAGS
+
+abinfo "Configuring binutils..."
+mkdir -p build
```
Absolute path here.

MingcongBai replied:
And verbosity.

MingcongBai commented on app-devel/binutils+cross-mips64r6el/autobuild/build:
```diff
@@ -0,0 +1,23 @@
+
+abinfo "Configuring binutils..."
+mkdir -p build
+cd build
```
Ditto.

MingcongBai commented on app-devel/binutils+cross-mips64r6el/autobuild/build:
```diff
@@ -0,0 +1,23 @@
+make
+
+abinfo "Installing binutils to target directory..."
+make DESTDIR=$PKGDIR install
```
Quote $PKGDIR.

MingcongBai commented on app-devel/gcc+cross-mips64r6el/autobuild/beyond:
```diff
@@ -0,0 +1,2 @@
+abinfo "Removing unnedded files..."
+rm $PKGDIR/opt/abcross/mips64r6el/share/info/dir
```
Verbosity here.

MingcongBai commented on app-devel/gcc+cross-mips64r6el/autobuild/build:
```diff
@@ -0,0 +1,36 @@
+unset CFLAGS CXXFLAGS LDFLAGS CPPFLAGS
+
+abinfo "Configuring GCC..."
+mkdir -p build
```
Ditto about absolute path.

MingcongBai replied:
And verbosity.

MingcongBai commented on app-devel/gcc+cross-mips64r6el/autobuild/build:
```diff
@@ -0,0 +1,36 @@
+
+abinfo "Configuring GCC..."
+mkdir -p build
+cd build
```
Ditto.

MingcongBai commented on app-devel/gcc+cross-mips64r6el/autobuild/build:
```diff
@@ -0,0 +1,36 @@
+     LD_FOR_TARGET=/opt/abcross/mips64r6el/bin/mipsisa64r6el-aosc-linux-gnuabi64-ld
+
+abinfo "Installing gcc into PKGDIR..."
+make DESTDIR=$PKGDIR install
```
Quote $PKGDIR.

Cyanoxygen committed:
binutils+cross-mips64r6el: lint script

- Quote unquoted variables.
- Replace more applicable strings with ${__CROSS}.
- Add missing verbosities.

Co-Authored-By: Mingcong Bai <jeffbai@aosc.xyz>

Cyanoxygen committed:
gcc+cross-mips64r6el: lint script

- Quote unquoted variables.
- Replace more applicable strings with ${__CROSS}.
- Add missing verbosities.

Co-Authored-By: Mingcong Bai <jeffbai@aosc.xyz>
